_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2045 @@
+
+Let's break this sentence down, starting at the end and working our way back:
+
+"... what's actually deployed"::
+  The only way to ensure that the infrastructure code in the `live` repository is an up-to-date representation of what's
+  actually deployed is to never make out-of-bound changes. As discussed in the previous section, out-of-bound changes
+  are sources of subtle bugs and undesirable actions in infrastructure code. If you get into a habit of making
+  out-of-bound changes, your infrastructure code will constantly need to make large amounts of changes to resolve the
+  configuration drift. This not only void many of the benefits of managing your infrastructure as code (e.g
+  reproducibility), it can also be a source of distrust in the code base. Note that the flip side is true as well: you
+  want to ensure that the code is continuously applied so that the code doesn't move too far ahead of the existing
+  infrastructure.
+
+"... a 1:1 representation ..."::
+  Keeping track of the infrastructure that has been deployed is a critical part of Site Reliability Engineering. If you
+  are not aware of all the environments that have been deployed, it is not only hard to make sense of any of the
+  monitoring alarms and metrics, it can also be a source of frustration when you need to debug an issue. Ensuring that
+  all the configuration for the environments are captured as code and live in your `live` repository provides an easy
+  and obvious way to know what has been deployed and the exact configuration in which the deployment happened. This not
+  only means avoiding out-of-bound changes, but also avoiding tooling where the configurations live outside the live
+  codebase (e.g using Terraform workspaces).
+
+"The master branch ..."::
+  You should only deploy your infrastructure from a single branch. This relates to the challenge of being able to only
+  have a single environment for the code. Since you can't have multiple deployed environments of the code, it becomes
+  hard to manage the shared infrastructure state across multiple copies of the code. This is most obvious with
+  Terraform and it's state tracking. Given the cost of spinning up multiple copies of your entire infrastructure, you
+  are typically forced to only maintain a single environment for your infrastructure code that everyone shares. This
+  means that applying the infrastructure from separate branches is the same as making out-of-band changes, because the
+  view is not unified. That is, terraform can only get the full view of the infrastructure if it merges all the changes
+  from the active branch together. In this fragmented view, there is a high likelihood that applying the infrastructure
+  in one branch can undo (as in, delete the infrastructure) the work of another branch because the configuration doesn't
+  include it. To avoid this, you need to have a single source of truth that is consistent. In trunk-based development,
+  that is the `master` branch.
```

Comment:
> Isn't it more accurate to say that you should only deploy your infrastructure from a tagged release? That is, all releases are cut after merging to the master branch, and you can run any number of environments using any number of releases.

I have to think through this again. FWIW I got this from https://blog.gruntwork.io/how-to-use-terraform-as-a-team-251bc1104973 (search "The golden rule"), and the section after describes the problem with branching.

That said, the more I think about it, the more I am beginning to realize that it might still make sense to only deploy from releases, even if that means releases are almost every commit.

---

> But there are multiple environments ... This is where I get lost.

I probably need a dedicated section for these points. Let me think about how best to describe this. I agree that it is very confusing, primarily because terms are overloaded (`environment`), and also because it is meta (`dev-dev` wtf!?).

But the gist is that you have to look at the infrastructure live code in its entirety, and consider what it means to have a `dev` environment for the live code in its entirety. The easiest way to see this is when you consider changes such as adding a completely new account. E.g you have a need for a `qa` account that is dedicated to your QA team. What would it mean to have a `dev` environment of that? What about when we are ready to deploy that? What would a release mean then? And should that release be different from what is already on the "prod" environment?

Slight detour: why would you even want a `dev` environment to test this in the first place? One possible failure mode is if you were hooking up VPC peering between the shared account `mgmt` VPC and `qa` environment, and you accidentally picked an overlapping VPC CIDR block for `qa` with `prod`, such that all your routes to `prod` broke and you lost the ability to VPN into `prod`.

The point I'm trying to make here (and in [the paragraph highlighted here](https://github.com/gruntwork-io/gruntwork-io.github.io/pull/265/files#r382259930)) is that in order to have a true dev environment for the live code, you need a complete replica of your entire AWS organization, where you can run apply on without breaking anything existing, but you can't have that.

Reply:
Completely revamped this section: cc17835